unistd/pwd.c

```diff
@@ -24,69 +25,92 @@
 
 struct passwd pwnam;
 char pw_name[NAME_MAX];
+char intstr[16];
 char pw_gecos[128];
 char pw_dir[PATH_MAX];
 char pw_shell[PATH_MAX];
 char pw_passwd[128];
 
 
-struct passwd *getpwnam(const char *name)
+static int readpwentry(FILE *fp, char *pwentry, unsigned int maxlgth)
 {
-	FILE *fp;
-	char buf[1024];
-
-	if ((fp = fopen(PASSWD_PATH, "r")) == NULL)
-		return NULL;
-
-	pwnam.pw_name = pw_name;
-	pwnam.pw_gecos = pw_name;
-	pwnam.pw_dir = pw_dir;
-	pwnam.pw_gecos = pw_gecos;
-	pwnam.pw_shell = pw_shell;
-	pwnam.pw_passwd = pw_passwd;
-
-	while (fgets(buf, sizeof(buf), fp)) {
-		sscanf (buf, "%[^:]:%[^:]:%d:%d:%[^:]:%[^:]:%s\n",
-			pwnam.pw_name, pwnam.pw_passwd,
-			&pwnam.pw_uid, &pwnam.pw_gid,
-			pwnam.pw_gecos, pwnam.pw_dir, pwnam.pw_shell);
-		if (strcmp(pwnam.pw_name, name) == 0) {
-			fclose(fp);
-			return &pwnam;
-		}
+	int c = 0;
+	unsigned int i = 0, isbad = 0;
+
+	c = fgetc(fp);
+	while (c != ':' && c != '\n' && c != EOF) {
+		if (i < maxlgth - 1)
+			pwentry[i++] = c;
+		else
+			isbad = 1;
+
+		c = fgetc(fp);
 	}
-	fclose(fp);
-	return NULL;
+	pwentry[i] = '\0';
+	return isbad;
 }
 
 
-struct passwd *getpwuid(uid_t uid)
+static inline int readuid(int readpwentry, uid_t *uid)
+{
+	int i = strlen(intstr);
+
+	*uid = strtoul(intstr, NULL, 10);
+	while (i > 0 && isdigit(intstr[--i]))
+		;
+	return (readpwentry && (++i));
```

Comment:
For MISRA fans out there:

```suggestion
	int i, err = 0;

	*uid = strtoul(intstr, NULL, 10);
	
	if (readpwentry != 0) {
		for (i = 0; instr[i] != '\0'; ++i) {
			if (!isdigit(instr[i])) {
				err = 1;
				break;
			}
		}
	}
	
	return err;
```

Reply:
in for a penny, in for a pound :) Added, thanks!